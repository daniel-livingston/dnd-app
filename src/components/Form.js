import React from "react";
import PropTypes from "prop-types";
import FormItem from "./FormItem";
import FormButton from "./FormButton";

const propTypes = {
	header: PropTypes.string,
	items: PropTypes.arrayOf(
		PropTypes.exact({
			title: PropTypes.string.isRequired,
			description: PropTypes.string,
			inputType: PropTypes.string.isRequired,
			inputValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
			inputOnChange: PropTypes.func.isRequired,
			inputPlaceholder: PropTypes.string.isRequired,
			selectOptions: PropTypes.arrayOf(
				PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired
			),
		})
	).isRequired,
	button: PropTypes.exact({
		onClick: PropTypes.func.isRequired,
		id: PropTypes.string,
		name: PropTypes.string.isRequired,
		justify: PropTypes.string,
		width: PropTypes.string,
	}),
};

const Form = (props) => (
	<form className='Form' onSubmit={(e) => e.preventDefault()}>
		{props.header && <h2 className='Form-header'>{props.header}</h2>}
		{props.items.map((item) => (
			<FormItem {...item} key={item.title} />
		))}
		{props.button && <FormButton {...props.button} />}
	</form>
);

Form.propTypes = propTypes;

export default Form;
