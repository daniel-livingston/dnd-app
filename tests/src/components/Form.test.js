import React from "react";
import { mount, shallow } from "enzyme";
import { Link, MemoryRouter } from "react-router-dom";
import Form from "../../../src/components/Form";
import { generateItem, generateForm } from "../../fixtures/Form";

// Possible other things to test:
// 1. Displaying headers and titles correctly
// 2. Changing class to reflect being a card or not

describe("submit options", () => {
	const textInput = generateItem({ type: "text", onChange: jest.fn() });
	const textInputForm = generateForm({
		items: [textInput],
		submit: {
			onSubmit: jest.fn(),
			buttonName: "Submit",
			buttonId: "test-button-id",
		},
	});

	test("renders button correctly", () => {
		const wrapper = shallow(<Form {...textInputForm} />);
		expect(wrapper).toMatchSnapshot();
	});

	test("button and form function correctly", () => {
		const wrapper = mount(<Form {...textInputForm} />);
		const button = wrapper.find("button");
		expect(button.prop("id")).toEqual(textInputForm.submit.buttonId);
		expect(button.text()).toEqual(textInputForm.submit.buttonName);
		wrapper.find("form").simulate("submit");
		expect(textInputForm.submit.onSubmit).toHaveBeenCalledTimes(1);
	});
});

// Input types
describe("link input", () => {
	const linkInput = generateItem({ type: "link", to: "/test", value: "Test Link" });
	const linkInputForm = generateForm({ items: [linkInput] });

	test("renders link input correctly", () => {
		const wrapper = shallow(<Form {...linkInputForm} />);
		expect(wrapper).toMatchSnapshot();
	});

	test("link options function correctly", () => {
		const wrapper = mount(
			<MemoryRouter>
				<Form {...linkInputForm} />
			</MemoryRouter>
		);
		const link = wrapper.find(Link);
		expect(link.prop("to")).toEqual(linkInput.to);
		expect(link.text()).toEqual(linkInput.value);
	});
});

describe("number input", () => {
	const numberInput = generateItem({ type: "number", onChange: jest.fn() });
	const numberInputForm = generateForm({ items: [numberInput] });

	test("renders number input correctly", () => {
		const wrapper = shallow(<Form {...numberInputForm} />);
		expect(wrapper).toMatchSnapshot();
	});

	test("number options function correctly", () => {
		const wrapper = mount(<Form {...numberInputForm} />);

		const number = wrapper.find("input");
		expect(number.prop("type")).toEqual(numberInput.type);
		expect(number.prop("value")).toEqual(numberInput.value);
		number.simulate("change");
		expect(numberInput.onChange).toHaveBeenCalledTimes(1);
	});
});

describe("password input", () => {
	const passwordInput = generateItem({ type: "password", onChange: jest.fn() });
	const passwordInputForm = generateForm({ items: [passwordInput] });

	test("renders text input correctly", () => {
		const wrapper = shallow(<Form {...passwordInputForm} />);
		expect(wrapper).toMatchSnapshot();
	});

	test("input options function correctly", () => {
		const wrapper = mount(<Form {...passwordInputForm} />);
		const input = wrapper.find("input");
		expect(input.prop("type")).toEqual(passwordInput.type);
		expect(input.prop("value")).toEqual(passwordInput.value);
		input.simulate("change");
		expect(passwordInput.onChange).toHaveBeenCalledTimes(1);
	});
});

describe("select input", () => {
	const selectInput = generateItem({
		type: "select",
		onChange: jest.fn(),
		value: "test1",
		options: ["test1", "test2", "test3"],
	});
	const selectForm = generateForm({ items: [selectInput] });

	test("renders select correctly", () => {
		const wrapper = shallow(<Form {...selectForm} />);
		expect(wrapper).toMatchSnapshot();
	});

	test("select options function correctly", () => {
		const wrapper = mount(<Form {...selectForm} />);
		const select = wrapper.find("select");
		const options = wrapper.find("option");
		expect(options).toHaveLength(selectInput.options.length);
		expect(select.prop("value")).toEqual(selectInput.value);
		select.simulate("change");
		expect(selectInput.onChange).toHaveBeenCalledTimes(1);
	});
});

describe("text input", () => {
	const textInput = generateItem({ type: "text", onChange: jest.fn() });
	const textInputForm = generateForm({ items: [textInput] });

	test("renders text input correctly", () => {
		const wrapper = shallow(<Form {...textInputForm} />);
		expect(wrapper).toMatchSnapshot();
	});

	test("input options function correctly", () => {
		const wrapper = mount(<Form {...textInputForm} />);
		const input = wrapper.find("input");
		expect(input.prop("type")).toEqual(textInput.type);
		expect(input.prop("value")).toEqual(textInput.value);
		input.simulate("change");
		expect(textInput.onChange).toHaveBeenCalledTimes(1);
	});
});
