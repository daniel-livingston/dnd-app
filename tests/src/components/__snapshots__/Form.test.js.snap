// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`link input renders link input correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form__section"
    key="Test header"
  >
    <h2
      className="form__section-header"
    >
      Test header
    </h2>
    <div
      className="form__section-body"
      key="Test title"
    >
      <div
        className="form__section-info"
      >
        <h3
          className="form__input-title"
        >
          Test title
        </h3>
        <p
          className="form__input-description"
        >
          Test description
        </p>
      </div>
      <Link
        className="form__link"
        to="/test"
      >
        Test Link
      </Link>
    </div>
  </div>
</form>
`;

exports[`number input renders number input correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form__section"
    key="Test header"
  >
    <h2
      className="form__section-header"
    >
      Test header
    </h2>
    <div
      className="form__section-body"
      key="Test title"
    >
      <div
        className="form__section-info"
      >
        <h3
          className="form__input-title"
        >
          Test title
        </h3>
        <p
          className="form__input-description"
        >
          Test description
        </p>
      </div>
      <input
        className="form__input"
        onChange={[MockFunction]}
        type="number"
        value=""
      />
    </div>
  </div>
</form>
`;

exports[`password input renders text input correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form__section"
    key="Test header"
  >
    <h2
      className="form__section-header"
    >
      Test header
    </h2>
    <div
      className="form__section-body"
      key="Test title"
    >
      <div
        className="form__section-info"
      >
        <h3
          className="form__input-title"
        >
          Test title
        </h3>
        <p
          className="form__input-description"
        >
          Test description
        </p>
      </div>
      <input
        className="form__input"
        onChange={[MockFunction]}
        placeholder="Test title"
        type="password"
        value=""
      />
    </div>
  </div>
</form>
`;

exports[`select input renders select correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form__section"
    key="Test header"
  >
    <h2
      className="form__section-header"
    >
      Test header
    </h2>
    <div
      className="form__section-body"
      key="Test title"
    >
      <div
        className="form__section-info"
      >
        <h3
          className="form__input-title"
        >
          Test title
        </h3>
        <p
          className="form__input-description"
        >
          Test description
        </p>
      </div>
      <select
        className="form__select"
        onChange={[MockFunction]}
        value="test1"
      >
        <option
          key="test1"
          value="test1"
        >
          test1
        </option>
        <option
          key="test2"
          value="test2"
        >
          test2
        </option>
        <option
          key="test3"
          value="test3"
        >
          test3
        </option>
      </select>
    </div>
  </div>
</form>
`;

exports[`submit options renders button correctly 1`] = `
<form
  className="form"
  onSubmit={[MockFunction]}
>
  <div
    className="form__section"
    key="Test header"
  >
    <h2
      className="form__section-header"
    >
      Test header
    </h2>
    <div
      className="form__section-body"
      key="Test title"
    >
      <div
        className="form__section-info"
      >
        <h3
          className="form__input-title"
        >
          Test title
        </h3>
        <p
          className="form__input-description"
        >
          Test description
        </p>
      </div>
      <input
        className="form__input"
        onChange={[MockFunction]}
        placeholder="Test title"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="form__button-container"
  >
    <button
      className="form__button"
      id="test-button-id"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`text input renders text input correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form__section"
    key="Test header"
  >
    <h2
      className="form__section-header"
    >
      Test header
    </h2>
    <div
      className="form__section-body"
      key="Test title"
    >
      <div
        className="form__section-info"
      >
        <h3
          className="form__input-title"
        >
          Test title
        </h3>
        <p
          className="form__input-description"
        >
          Test description
        </p>
      </div>
      <input
        className="form__input"
        onChange={[MockFunction]}
        placeholder="Test title"
        type="text"
        value=""
      />
    </div>
  </div>
</form>
`;
